### ISSUE TITLE
Custom Network with other caffe build

### ISSUE THREAD

**Issue by Almo1982** [NVIDIA/DIGITS#882]
Hi all,

did anybody succesfully trained custom networks that include different layers that the official caffe-build with NVIDIA DIGITS?

As far as i understand only the "official" BVLC/caffe is supported from DIGITS but isn't there any way to use other caffe-builds within DIGITS?

Actually the error i get is that when i copy the train net-definition into the custom Network area and try to start training it says "no Name bn_param in caffe.Layerparam". I recognized that there is no bn_param.cpp in the Layer Folder of the official caffe build, what explains this error. In other Forum i read that caffe supports "Batch_norm_param"-layer, but if i replace bn_param with Batch_norm_param another error occurs "no Name shift_filler found". 

Is there any Chance to get it work by e.g. copying the missing (bn_param) layer into the Folders of the official caffe build? i guess ist not that easy, i tried a lot of things but nothing worked so far.

Or is there a possibility to tell DIGITS not to use the "Standard" caffe-build but the "custom-caffe"-build? I added the custom-caffe build path to my bashrc, didnt work too...

Any help would be appreciated!

thank you.

Regards
 Stefan

**Comment by douzsh** [comment]
You should modify your caffe.proto and get the param definition. 
I think you should merge the caffe.proto in NVcaffe with yourcaffe.
After this, you can add your personal layers and get a work version.

**Comment by Almo1982** [comment]
Thank you very much for your Response!

i am not very familiar with changing code inside the caffe structure, could you give me a short tutorial how to proceed with this?

Thank you in advance!

**Comment by douzsh** [comment]
I don't think I can help you too much because there is no tutorial. Every state of art cnn model defined many new layers and new params. I think you can follow this steps to make the changes.
1. merge the caffe.proto to make sure every parameter is defined in final version.
2. copy the new layers and related util files to the final version.
3. merge the util files both used.
This is always not a easy job. If you want to merge branches based on BLVC/caffe with NVIDIA/caffe you should also pay attentions to the parelle GPU issues.
Anyway, you should familiar with caffe at first.

**Comment by Almo1982** [comment]
Hi douzsh,

ok, thank you anyway for your help, i will try to figure this out.
I already trained some Networks with custom caffe-builds sucessfully, the thing is just, i want to visualize the Training process and the convolutional layers and this works great with DIGITS (at least for the Standard Networks that worked for me so far).

**Comment by Almo1982** [comment]
hmm, i didnt make any success so far after hours of trying out to merge the custom data...

another question, where can i see the DIGITS path to the caffe-version that is used? its kind of strange, any changes i make have absolutely no influence at all to DIGITS?! 

Stefan

**Comment by douzsh** [comment]
I think you should first train your model by command line like 
./build/tools/caffe train --solver='solver.pt'
If you can train your model sucessful with nvcaffe, than you can try 
export CAFFE_ROOT=/path/to/caffe/
and restart your digits server.

**Comment by Almo1982** [comment]
Thank you douzsh!

What is not really clear to me, is DIGITS using the "normal" caffe build i cloned and built from BVLC\caffe? For my understanding there exists a NVIDIA\caffe build, but i cant find this Installation at all on my Computer. I installed DIGITS by the Installation guide under https://github.com/NVIDIA/DIGITS/blob/master/docs/UbuntuInstall.md 
, i didnt had to define any caffe path or anything else. so i am a Little bit confused what caffe built DIGITS actually uses...

my idea was to Change the caffe path for DIGITS to my custom caffe build...but it seems not to be that easy?! On the other Hand, why shouldnt it work if i tell DIGITS the custom build caffe path  and DIGITS gets the (new) caffe layers etc. from this source...this is actually my understanding Problem.

regards
Stefan

**Comment by douzsh** [comment]
I didnot use apt-get for digits installation.
I can answer part of your questions.
1. digits use nvidia/caffe but not BVLC/caffe.
2. If you want to change config settings of your apt-get digits, you can find command on your link page. sudo python -m digits.config.edit -v
    with which you can change the config file and update to digits server.

**Comment by gheinrich** [comment]
Going back to the initial comment the issue is @Almo1982 is using an older specification for network description. The specification for nv-caffe 0.14 may be found [there](https://github.com/NVIDIA/caffe/blob/caffe-0.14/src/caffe/proto/caffe.proto) this is mostly in line with the latest spec from the BVLC tree. As @Almo1982 correctly found out `bn_param` has been renamed `batch_norm_param` with this [commit from BVLC fork](https://github.com/NVIDIA/caffe/commit/e48dcda9c46426f8d5cd568b791d09e68add427b). Also note that `shift_filler` was removed.

In summary: @Almo1982 wherever your custom network is coming from it is using an old format. You need to upgrade to a more recent spec.

**Comment by Almo1982** [comment]
Thank you for this Information!

Slowly i think i understand the structure of the caffe build...that means i have to rework my old custom Network in my Train_val.prototxt to fit the new (current) caffe spec, e.g. changing the bn_param to the actual valid batch_norm_param declaration. When shift_filler is removed, i am asking myself how to replace this then?! Just erasing this cant be the solution??? So there is a last understanding Problem...how to replace old declarations (bn_param) with new ones (Batch_norm_param) that have different Parameter values?

I have to work through this in very detail, i dont see any other Chance ;-)

**Comment by gheinrich** [comment]
I think the [Caffe users list](https://groups.google.com/forum/#!forum/caffe-users) is a better place to ask those questions.
